File: nobodywho/python/tests/test_tool_calling.py
import os

import nobodywho
import pytest

import logging

logging.addLevelName(5, "TRACE")


def get_tool_calls(chat_history):
    """Extract all tool calls from chat history."""
    tool_calls = []
    for msg in chat_history:
        if "tool_calls" in msg:
            tool_calls.extend(msg["tool_calls"])
    return tool_calls


def get_tool_responses(chat_history):
    """Extract all tool responses from chat history."""
    return [msg for msg in chat_history if msg.get("role") == "tool"]

@nobodywho.tool(
    description="Boop foob",
    params={
        "reflarb": "the clump factor for the flopar",
        "unfloop": "activate the rotational velocidensity collider",
    },
)
def reflarbicator(reflarb: int, unfloop: bool) -> str:
    return "hahaha"


@nobodywho.tool(
    description="Gets the weather for a given location",
    params={"location": "The city or location to get weather for"}
)
def get_weather(location: str) -> str:
    return f"The weather in {location} is sunny and 21°C"


@nobodywho.tool(
    description="Applies the sparklify effect to a given piece of text."
)
async def async_sparklify(text: str) -> str:
    return f"✨{text.upper()}✨"



@nobodywho.tool(description="Applies the sparklify effect to a given piece of text.")
def sparklify(text: str) -> str:
    return f"✨{text.upper()}✨"


@nobodywho.tool(description="Returns the intersection of set1 and set2")
def set_intersection(set1 : set[int], set2 : set[int]) -> str:
    return str(set1.intersection(set2))

@nobodywho.tool(description="Returns the string in the tuple string_int_pair concatenated with itself a number of times equal to the integer in string_int_pair")
def multiply_strings(string_int_pair : tuple[str,int]) -> str:
    assert isinstance(string_int_pair, tuple)
    return string_int_pair[0] * string_int_pair[1]

@nobodywho.tool(description="Does vector addition on the list of vectors provided", params={"list_of_vectors" : "List of vectors to added. Each vector must of same length"})
def add_list_of_vectors(list_of_vectors : list[list[int]]) -> str:
    res = list_of_vectors[0].copy()
    for v in list_of_vectors[1:]:
        for i,coord in enumerate(v):
            res[i] += coord
    return str(res)

@nobodywho.tool(description="Returns the volume of a cube with the input dimenensions", params={"dimensions" : "A map containing the numeric values for the width, heigth and depth of a cube."})
def calculate_volume(dimensions: dict[str,float]) -> str:
      # dimensions: {"width": float, "height": float, "depth": float}
      return str(dimensions["width"] * dimensions["height"] * dimensions["depth"])


@pytest.fixture(scope="module")
def model():
    model_path = os.environ.get("TEST_MODEL")
    if not model_path:
        raise ValueError("TEST_MODEL environment variable is not set")

    return nobodywho.Model(model_path)

@pytest.fixture
def chat(model):
    if "qwen" in os.environ.get("TEST_MODEL", "").lower():
        return nobodywho.Chat(
            model, system_prompt="You are a helpful assistant", allow_thinking=False, tools=[set_intersection, multiply_strings, sparklify, reflarbicator, add_list_of_vectors, calculate_volume]
        )
    return nobodywho.Chat(
            model, system_prompt="You are a helpful assistant", allow_thinking=False, tools=[sparklify]
        )
    



def test_tool_construction():
    assert sparklify is not None
    assert isinstance(sparklify, nobodywho.Tool)
    assert sparklify("foobar") == "✨FOOBAR✨"


def test_tool_calling(chat):
    chat.ask("Please sparklify this word: 'julemand' and show me the result").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "sparklify"
    assert tool_calls[0]["function"]["arguments"]["text"] == "julemand"

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "sparklify"
    assert tool_responses[0]["content"] == "✨JULEMAND✨"




def test_tool_bad_parameters():
    with pytest.raises(TypeError):

        @nobodywho.tool(description="foobar", params={"b": "uh-oh"})
        def i_fucked_up(a: int) -> str:
            return "fuck"


@pytest.mark.asyncio
async def test_async_tool_construction():
    assert async_sparklify is not None
    assert isinstance(async_sparklify, nobodywho.Tool)
    assert await async_sparklify("foobar") == "✨FOOBAR✨"


def test_async_tool_calling(model):
    chat = nobodywho.Chat(model, tools=[async_sparklify])
    chat.ask("Please sparklify this word: 'julemand' and show me the result").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "async_sparklify"
    assert tool_calls[0]["function"]["arguments"]["text"] == "julemand"

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "async_sparklify"
    assert tool_responses[0]["content"] == "✨JULEMAND✨"


def test_async_tool_bad_parameters():
    with pytest.raises(TypeError):

        @nobodywho.tool(description="foobar", params={"b": "uh-oh"})
        async def i_fucked_up(a: int) -> str:
            return "fuck"



def test_set_tools(model):
    """Test that set_tools changes available tools and persists after reset_history"""
    chat = nobodywho.Chat(
            model, system_prompt="You are a helpful assistant", allow_thinking=False, tools=[sparklify]
        )
    # Use initial tool
    chat.ask("Please sparklify this word: 'julemand' and show me the result").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "sparklify"
    assert tool_calls[0]["function"]["arguments"]["text"] == "julemand"

    assert len(tool_responses) == 1
    assert tool_responses[0]["content"] == "✨JULEMAND✨"

    # Change tools
    chat.set_tools([get_weather])

    # Clear history but keep new tools
    chat.reset_history()

    # Try to use new tool - should work
    chat.ask("What's the weather in Copenhagen?").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "get_weather"
    assert tool_calls[0]["function"]["arguments"]["location"] == "Copenhagen"

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "get_weather"
    assert tool_responses[0]["content"] == "The weather in Copenhagen is sunny and 21°C"


def test_tool_calling_with_custom_sampler(model):
    chat = nobodywho.Chat(
        model,
        tools=[sparklify],
        sampler=nobodywho.SamplerBuilder()
            .top_k(64)
            .top_p(0.95, min_keep=2)
            .temperature(0.8)
            .dist(),
        allow_thinking=False,
    )

    chat.ask("Please sparklify this word: 'julemand' and show me the result").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "sparklify"
    assert tool_calls[0]["function"]["arguments"]["text"] == "julemand"

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "sparklify"
    assert tool_responses[0]["content"] == "✨JULEMAND✨"


def test_tool_with_sets(chat):
    if "qwen" not in os.environ.get("TEST_MODEL", "").lower():
        pytest.skip("Test only runs with Qwen models")

    chat.ask("Please use the provided tool to find the intersection between the sets {12,5,7,3,4} and {12,9,5,3}").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "set_intersection"
    assert set(tool_calls[0]["function"]["arguments"]["set1"]) == {12, 5, 7, 3, 4}
    assert set(tool_calls[0]["function"]["arguments"]["set2"]) == {12, 9, 5, 3}

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "set_intersection"
    # The response is a string representation of a set, check all expected elements are present
    response_content = tool_responses[0]["content"]
    assert "12" in response_content and "5" in response_content and "3" in response_content 

def test_tool_with_tuple(chat):
    if "qwen" not in os.environ.get("TEST_MODEL", "").lower():
        pytest.skip("Test only runs with Qwen models")

    chat.ask("Please use the provided tool to multiply the string BingBong by 3").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "multiply_strings"
    assert tool_calls[0]["function"]["arguments"]["string_int_pair"] == ["BingBong", 3]

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "multiply_strings"
    assert tool_responses[0]["content"] == "BingBongBingBongBingBong"

def test_tool_with_nested_list(chat):
    if "qwen" not in os.environ.get("TEST_MODEL", "").lower():
        pytest.skip("Test only runs with Qwen models")

    chat.ask("Please use the provided tool to add the vectors [[1,2,3],[4,5,6],[7,8,9]].").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "add_list_of_vectors"
    assert tool_calls[0]["function"]["arguments"]["list_of_vectors"] == [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "add_list_of_vectors"
    assert tool_responses[0]["content"] == "[12, 15, 18]"

def test_tool_with_dict(chat):
    if "qwen" not in os.environ.get("TEST_MODEL", "").lower():
        pytest.skip("Test only runs with Qwen models")

    chat.ask("Please use the provided tool to find the volume of a cube with dimensions 30 x 20 x 10.").completed()

    history = chat.get_chat_history()
    tool_calls = get_tool_calls(history)
    tool_responses = get_tool_responses(history)

    assert len(tool_calls) == 1
    assert tool_calls[0]["function"]["name"] == "calculate_volume"
    dimensions = tool_calls[0]["function"]["arguments"]["dimensions"]
    assert dimensions["width"] == 30
    assert dimensions["height"] == 20
    assert dimensions["depth"] == 10

    assert len(tool_responses) == 1
    assert tool_responses[0]["name"] == "calculate_volume"
    assert tool_responses[0]["content"] == "6000.0"

File: nobodywho/rust/tests/test_tool_calling.rs
use nobodywho_rust::{Chat, Message, Model, SamplerBuilder, Tool, ToolCall};
use std::sync::Arc;

fn model_path() -> String {
    std::env::var("TEST_MODEL").unwrap_or_else(|_| "model.gguf".to_string())
}

fn load_model() -> Model {
    Model::builder(model_path())
        .build()
        .expect("failed to load model")
}

fn make_chat(model: &Model, tools: Vec<Tool>) -> Chat {
    Chat::builder(model)
        .with_system_prompt("You are a helpful assistant.")
        .with_template_variable("enable_thinking", false)
        .with_tools(tools)
        .build()
}

fn sparklify_tool() -> Tool {
    let schema = serde_json::json!({
        "type": "object",
        "properties": {
            "text": {"type": "string", "description": "The text to sparklify"}
        },
        "required": ["text"],
        "additionalProperties": false
    });
    Tool::new(
        "sparklify".to_string(),
        "Applies the sparklify effect to a given piece of text.".to_string(),
        schema,
        Arc::new(|args: serde_json::Value| {
            let text = args["text"].as_str().unwrap_or("");
            format!("✨{}✨", text.to_uppercase())
        }),
    )
}

fn get_weather_tool() -> Tool {
    let schema = serde_json::json!({
        "type": "object",
        "properties": {
            "location": {"type": "string", "description": "The city or location to get weather for"}
        },
        "required": ["location"],
        "additionalProperties": false
    });
    Tool::new(
        "get_weather".to_string(),
        "Gets the weather for a given location.".to_string(),
        schema,
        Arc::new(|args: serde_json::Value| {
            let location = args["location"].as_str().unwrap_or("unknown");
            format!("The weather in {location} is sunny and 21°C")
        }),
    )
}

fn get_tool_calls(history: &[Message]) -> Vec<ToolCall> {
    history
        .iter()
        .flat_map(|m| match m {
            Message::ToolCalls { tool_calls, .. } => tool_calls.clone(),
            _ => vec![],
        })
        .collect()
}

fn get_tool_responses(history: &[Message]) -> Vec<&Message> {
    history
        .iter()
        .filter(|m| matches!(m, Message::ToolResp { .. }))
        .collect()
}

// ── Tool construction ─────────────────────────────────────────────────────────

#[test]
fn test_tool_construction() {
    let tool = sparklify_tool();
    // invoke the tool directly to verify the callback works
    let result = (tool.function)(serde_json::json!({"text": "foobar"}));
    assert_eq!(result, "✨FOOBAR✨");
}

// ── Tool calling ──────────────────────────────────────────────────────────────

#[test]
fn test_tool_calling() {
    let model = load_model();
    let chat = make_chat(&model, vec![sparklify_tool()]);

    chat.ask("Please sparklify this word: 'julemand' and show me the result")
        .completed()
        .unwrap();

    let history = chat.get_chat_history().unwrap();
    let tool_calls = get_tool_calls(&history);
    let tool_responses = get_tool_responses(&history);

    assert_eq!(tool_calls.len(), 1);
    assert_eq!(tool_calls[0].name, "sparklify");
    assert_eq!(tool_calls[0].arguments["text"], "julemand");

    assert_eq!(tool_responses.len(), 1);
    if let Message::ToolResp { name, content, .. } = tool_responses[0] {
        assert_eq!(name, "sparklify");
        assert_eq!(content, "✨JULEMAND✨");
    }
}

// ── set_tools ─────────────────────────────────────────────────────────────────

#[test]
fn test_set_tools() {
    let model = load_model();
    let chat = make_chat(&model, vec![sparklify_tool()]);

    chat.ask("Please sparklify this word: 'julemand' and show me the result")
        .completed()
        .unwrap();

    let history = chat.get_chat_history().unwrap();
    let tool_calls = get_tool_calls(&history);
    assert_eq!(tool_calls.len(), 1);
    assert_eq!(tool_calls[0].name, "sparklify");

    // swap to a different tool and clear history
    chat.set_tools(vec![get_weather_tool()]).unwrap();
    chat.reset_history().unwrap();

    chat.ask("What's the weather in Copenhagen?")
        .completed()
        .unwrap();

    let history = chat.get_chat_history().unwrap();
    let tool_calls = get_tool_calls(&history);
    let tool_responses = get_tool_responses(&history);

    assert_eq!(tool_calls.len(), 1);
    assert_eq!(tool_calls[0].name, "get_weather");
    assert_eq!(tool_calls[0].arguments["location"], "Copenhagen");

    assert_eq!(tool_responses.len(), 1);
    if let Message::ToolResp { content, .. } = tool_responses[0] {
        assert_eq!(content, "The weather in Copenhagen is sunny and 21°C");
    }
}

// ── Custom sampler ────────────────────────────────────────────────────────────

#[test]
fn test_tool_calling_with_custom_sampler() {
    let model = load_model();
    let sampler = SamplerBuilder::new()
        .top_k(64)
        .top_p(0.95, 2)
        .temperature(0.8)
        .dist();

    let chat = Chat::builder(&model)
        .with_system_prompt("You are a helpful assistant.")
        .with_template_variable("enable_thinking", false)
        .with_tool(sparklify_tool())
        .with_sampler(sampler)
        .build();

    chat.ask("Please sparklify this word: 'julemand' and show me the result")
        .completed()
        .unwrap();

    let history = chat.get_chat_history().unwrap();
    let tool_calls = get_tool_calls(&history);
    let tool_responses = get_tool_responses(&history);

    assert_eq!(tool_calls.len(), 1);
    assert_eq!(tool_calls[0].name, "sparklify");
    assert_eq!(tool_calls[0].arguments["text"], "julemand");

    assert_eq!(tool_responses.len(), 1);
    if let Message::ToolResp { content, .. } = tool_responses[0] {
        assert_eq!(content, "✨JULEMAND✨");
    }
}
